### DISCUSSION TITLE
[feature or bug] cargo build failed by dynamic link

### DISCUSSION THREAD

**Discussion by SoTosorrow** [bevyengine/bevy#5354] [Q&A]
## Bevy version
version = "0.7.0", features = ["dynamic"]

## Operating system
ubuntu 20.04

## What problem does this solve or what need does it fill?
```
cargo build
./target/debug/game: error while loading shared libraries: libbevy_dylib-240b7c17efd23e2e.so: cannot open shared object file: No such file or directory
 ```
Hello ,thanks for the bevy first.

The question is: After I cargo build, I got a libbevy_dylib.so and a execute file. but the execute file need a libbbevy_dylib-xxxxx.so. 
I also have a libbbevy_dylib-xxxxx.so in "target/debug/deps" but it also can not be used for the execute file.

So Is there dynamic feature still a temporary function just for faster project compile but can not be used in release?
There is a similar issue #2051 report on 30 Apr 2021. After a year, it still looks like we haven't updated the feature on stable?

If it wasn't for my operational error, maybe we should have written it clearly in the documentation or bevy book.
Maybe it's already in your plans, Anyway, I think this is a feature worth adding.

**Comment by Nilirad** [comment]
Converting to a discussion since it is known that the dynamic feature normally works on Linux distros...

EDIT: let's wait , just saw @alice-i-cecile labeling 😄

**Reply by alice-i-cecile** [reply]
Discussions can be labelled too!

**Reply by alice-i-cecile** [reply]
But yes, this is more useful as a discussion as this problem is almost certainly setup specific.

**Reply by SoTosorrow** [reply]
thanks for your convert! Should it add a label of compile failure? actually the cargo build is success but the link is failed.

**Reply by alice-i-cecile** [reply]
Most users searching won't have that level of sophistication: it'll look like a compile failure. So I think we keep the label :)

**Reply by SoTosorrow** [reply]
Ok! thx for your reply

**Comment by Nilirad** [comment]
> So Is there dynamic feature still a temporary function just for faster project compile but can not be used in release?

Technically, you can, but you'll also have to ship the Bevy dynamic library and the std dynamic library, negating the benefits of shared libraries.

Also, you can't even share the same Bevy library for many released games since Rust does not guarantee ABI stability across compiler versions/snapshots.

**Reply by SoTosorrow** [reply]
> Also, you can't even share the same Bevy library for many released games since Rust does not guarantee ABI stability across compiler versions/snapshots.

Oh I see! but the .so which cargo build provided is also can't be used for myself?

**Reply by Nilirad** [reply]
Technically you probably could reuse the .so library, but it is impractical since cargo doesn't support building on precompiled dependencies, so you would have to rely on lower level tools.

**Comment by AnotherCoder1** [comment]
I have the same problem with you. But in my Cargo.toml file look like this `[dependencies]\nbevy={version="0.11", features="dynamic_linking"}` instead of "dynamic" feature as you put in. As for now, after the first long compilation process by issuing "cargo build", I use "cargo run" to technically recompile again but this time it is faster, after that it runs the program automatically. But, If i tried to run/execute the executable manually in the target/debug directory it cannot find "libbevy_dylib-xxx.so" like your case. After searching in the deps folder, I have found the `.so` file in there. Tried to copy it outside the deps i.e into target/debug i got different .so file not found "libstd-xxx.so". Tried to ask AI, it says it is a rust standard library. Also tried to add the folder directory in the user envirionment variable in the .bashrc file using the name LD_LIBRARY_PATH but no luck. At the end just resort to "cargo run" which is essentially a recompilation.

**Reply by tim-blackbird** [reply]
`cargo run` only re-compiles when necessary, other than that it's the same as running the binary.

When compiling an executable meant to be used on its own the "dynamic_linking" feature should be disabled.

**Reply by AnotherCoder1** [reply]
finally, I manage to run the program manually without cargo. With the help of AI the "libstd-xxx.so" file is located in the directory where rustup was installed, in my case in ubuntu it is located in `.rustup/toolchains/..`. After copying the .so file into `target/debug`, in that directory open the terminal and run the executable using the command `./<executable_name>` and it run as usual without using cargo.

Disabling `dynamic_linking` is not a choice for me as I do not want to recompile bevy every time I want to start a new project. This is not the optimal way, but at least I know now how to run it manually without cargo.

**Reply by p0ryae** [reply]
@AnotherCoder1 Did you stumble upon an error relating `libsysinfo` rlib after fixing the issue with `libbevy_dylib-xxx.so`?

**Reply by AnotherCoder1** [reply]
No.
